src/mlj_interface.jl

```diff
@@ -0,0 +1,170 @@
+# TODO 1: a using MLJModelInterface or import MLJModelInterface statement
+# Expose all instances of user specified structs and package artifcats.
+const ParallelKMeans_Desc = "Parallel & lightning fast implementation of all variants of the KMeans clustering algorithm in native Julia."
+
+# availalbe variants for reference
+const MLJDICT = Dict(:Lloyd => Lloyd(),
+                     :Hamerly => Hamerly(),
+                     :LightElkan => LightElkan())
+
+####
+#### MODEL DEFINITION
+####
+# TODO 2: MLJ-compatible model types and constructors
+
+mutable struct KMeans <: MLJModelInterface.Unsupervised
+    algo::Symbol
+    k_init::String
+    k::Int
+    tol::Float64
+    max_iters::Int
+    copy::Bool
+    threads::Int
+    verbosity::Int
+    init
+end
+
+
+function KMeans(; algo=:Lloyd, k_init="k-means++",
+                k=3, tol=1e-6, max_iters=300, copy=true,
+                threads=Threads.nthreads(), verbosity=0, init=nothing)
+
+    model   = KMeans(algo, k_init, k, tol, max_iters, copy, threads, verbosity, init)
+    message = MLJModelInterface.clean!(model)
+    isempty(message) || @warn message
+    return model
+end
+
+
+function MLJModelInterface.clean!(m::KMeans)
+    warning = ""
+
+    if !(m.algo ∈ keys(MLJDICT))
+        warning *= "Unsuppored algorithm supplied. Defauting to KMeans++ seeding algorithm."
+        m.algo = :Lloyd
+
+    elseif m.k_init != "k-means++"
+        warning *= "Only `k-means++` or random seeding algorithms are supported. Defaulting to random seeding."
+        m.k_init = "random"
+
+    elseif m.k < 1
+        warning *= "Number of clusters must be greater than 0. Defaulting to 3 clusters."
+        m.k = 3
+
+    elseif !(m.tol < 1.0)
+        warning *= "Tolerance level must be less than 1. Defaulting to tol of 1e-6."
+        m.tol = 1e-6
+
+    elseif !(m.max_iters > 0)
+        warning *= "Number of permitted iterations must be greater than 0. Defaulting to 300 iterations."
+        m.max_iters = 300
+
+    elseif !(m.threads > 0)
+        warning *= "Number of threads must be at least 1. Defaulting to all threads available."
+        m.threads = Threads.nthreads()
```

Comment:
Shouldn't you add a warning if the number of threads is larger than the overall number of threads?

Reply:
> Shouldn't you add a warning if the number of threads is larger than the overall number of threads?

Agreed.

Reply:
Actually not.
"Number of threads" here is actually a number of chunks that will be used for processing. It can be larger than the number of threads, and it's safe to use, but it wouldn't give performance boost.

Reply:
I will keep it as it for now. Until we have tested any potential benefits of more threads than available cores.